sdk/nodejs/tsutils.ts

```diff
@@ -12,22 +12,56 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import * as fs from "fs";
+// The typescript and tsnode imports are used for type-checking only. Do not reference them in the emitted code.
 import * as typescript from "typescript";
-
+import * as tsnode from "ts-node";
+import * as fs from "fs";
+import * as path from "path";
 import * as log from "./log";
 
 /** @internal */
 export function loadTypeScriptCompilerOptions(tsConfigPath: string): object {
     try {
         const tsConfigString = fs.readFileSync(tsConfigPath).toString();
-        // Using local `require("typescript")` to avoid always loading
+        // Using local `require("./typescript-shim")` to avoid always loading
         // and only load on-demand, avoid up to 300s overhead in Node runtime.
-        const ts: typeof typescript = require("typescript");
+        const ts: typeof typescript = require("./typescript-shim");
         const tsConfig = ts.parseConfigFileTextToJson(tsConfigPath, tsConfigString).config;
         return tsConfig["compilerOptions"] ?? {};
     } catch (err) {
         log.debug(`Ignoring error in loadCompilerOptions(${tsConfigPath}}): ${err}`);
         return {};
     }
 }
+
+/**
+ * @internal Determine the strings used for requiring typescript and ts-node.
+ */
+export function typeScriptRequireStrings(): { typescriptRequire: string; tsnodeRequire: string } {
+    let typescriptRequire = "typescript";
+    let tsnodeRequire = "ts-node";
+
+    try {
+        // Try loading typescript from node_modules
+        const ts: typeof typescript = require("typescript");
+        const tsPath = require.resolve("typescript");
+        log.debug(`Found typescript version ${ts.version} at ${tsPath}`);
+    } catch (error) {
+        // Fallback to the vendored version
+        typescriptRequire = path.join(__dirname, "vendor", "typescript@3.8.3", "typescript.js");
```

Comment:
Woukd it be worth having a `log.debug` for when we're falling back to the vendored version (here and below)?